### PR TITLE
Implement b2 version of encoder

### PR DESCRIPTION
**Pull request by hajimehoshi**

### CODE REVIEW

File: go/signedexchange/signedexchange.go [hunk 1]
```diff
@@ -95,6 +107,7 @@ func (e *Exchange) encodeRequestCommon(enc *cbor.Encoder) []*cbor.MapEntryEncode
 			keyE.EncodeByteString(keyMethod)
 			valueE.EncodeByteString(valueGet)
 		}),
+		// TODO: Don't encode URL when the version is not 1b1
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply go/signedexchange/signedexchange.go]: Oops, nice catch

File: go/signedexchange/cmd/gen-signedexchange/main.go [hunk 2]
```diff
@@ -30,6 +31,7 @@ func (h *headerArgs) Set(value string) error {
 
 var (
 	flagUri            = flag.String("uri", "https://example.com/index.html", "The URI of the resource represented in the exchange")
+	flagVersion        = flag.String("version", "1b2", "The signedexchange version")
```
Comment (irori) [inline go/signedexchange/cmd/gen-signedexchange/main.go]: Let's use "1b1" as default until we have all the b2 changes in (i.e. http-mice-03).

File: go/signedexchange/signedexchange.go [hunk 3]
```diff
@@ -14,8 +14,22 @@ import (
...
 )
 
+const HeaderMagicBytesLen = 8
+
+func HeaderMagicBytes(v version.Version) []byte {
+	switch v {
+	case version.Version1b1:
+		return []byte("sxg1-b1\x00")
+	case version.Version1b2:
+		return []byte("sxg1\x00\x00\x00\x00")
```
Comment (nyaxt) [inline go/signedexchange/signedexchange.go]: This should be `"sxg1-b2\x00"`

Comment (hajimehoshi) [reply go/signedexchange/signedexchange.go]: I thought the spec said the magic header byte is `"sxg1\x00\x00\x00\x00"`.

> 8 bytes consisting of the ASCII characters “sxg1” followed by 4 0x00 bytes, to serve as a file signature. This is redundant with the MIME type, and recipients that receive both MUST check that they match and stop parsing if they don’t.

File: go/signedexchange/cmd/gen-signedexchange/main.go (hunk 2 above)
Comment (hajimehoshi) [reply go/signedexchange/cmd/gen-signedexchange/main.go]: Agreed. Change this later then.

File: go/signedexchange/signedexchange.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply go/signedexchange/signedexchange.go]: Done.

File: go/signedexchange/signedexchange.go [hunk 4]
```diff
@@ -230,53 +243,124 @@ func (e *Exchange) decodeExchangeHeaders(dec *cbor.Decoder) error {
...
+		}
+
+		// "3. fallbackUrlLength bytes holding a fallbackUrl, which MUST be an absolute URL with a scheme of “https”.
+		// Note: The byte location of the fallback URL is intended to remain invariant across versions of the application/signed-exchange format so that parsers encountering unknown versions can always find a URL to redirect to." [spec text]
+		if _, err := w.Write([]byte(url)); err != nil {
+			return err
+		}
+
+		// "4. 3 bytes storing a big-endian integer sigLength. If this is larger than 16384 (16*1024), parsing MUST fail." [spec text]
+		if len(e.SignatureHeaderValue) > 16*1024 {
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply go/signedexchange/signedexchange.go]: Done (defined a local unexported const)

File: go/signedexchange/signedexchange.go [hunk 5]
```diff
@@ -230,53 +243,124 @@ func (e *Exchange) decodeExchangeHeaders(dec *cbor.Decoder) error {
...
+		encodedSigLength, err := bigendian.EncodeBytesUint(int64(len(e.SignatureHeaderValue)), 3)
+		if err != nil {
+			return err
+		}
+		if _, err := w.Write(encodedSigLength); err != nil {
+			return err
+		}
+
+		// "5. 3 bytes storing a big-endian integer headerLength. If this is larger than 524288 (512*1024), parsing MUST fail." [spec text]
+		if headerLength > 512*1024 {
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply go/signedexchange/signedexchange.go]: Done

File: go/signedexchange/signedexchange.go (hunk 3 above)
Comment (hajimehoshi) [reply go/signedexchange/signedexchange.go]: Done, but I'd like to know if the spec is correct or not.

**Pull request merged**